From 9071ac2e2153e811358842c6d7e5eb5e34f2ffe2 Mon Sep 17 00:00:00 2001
From: Min Fanlei <minfanlei@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:09 +0800
Subject: [PATCH] sw64: enable more than 32 CPUs for guest
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56WV8

--------------------------------

Add chip3_setup_vt_core_start() implementation to support more than
32 CPUs for guest os.

Signed-off-by: Min Fanlei <minfanlei@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/chip/chip3/chip.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/arch/sw_64/chip/chip3/chip.c b/arch/sw_64/chip/chip3/chip.c
index f4b17cfa1127..fcd53775a9c9 100644
--- a/arch/sw_64/chip/chip3/chip.c
+++ b/arch/sw_64/chip/chip3/chip.c
@@ -133,6 +133,19 @@ static unsigned long chip3_get_node_mem(int nodeid)
 	return node_mem;
 }
 
+static void chip3_setup_vt_core_start(struct cpumask *cpumask)
+{
+	int i;
+	unsigned long coreonline;
+
+	coreonline = sw64_io_read(0, CORE_ONLINE);
+
+	for (i = 0; i < 64 ; i++) {
+		if (coreonline & (1UL << i))
+			cpumask_set_cpu(i, cpumask);
+	}
+}
+
 static void chip3_setup_core_start(struct cpumask *cpumask)
 {
 	int i, j, cpus;
@@ -574,6 +587,7 @@ static void chip3_hose_init(struct pci_controller *hose)
 static void chip3_init_ops_fixup(void)
 {
 	if (is_guest_or_emul()) {
+		sw64_chip_init->early_init.setup_core_start = chip3_setup_vt_core_start;
 		sw64_chip_init->early_init.get_node_mem = chip3_get_vt_node_mem;
 		sw64_chip_init->pci_init.check_pci_linkup = chip3_check_pci_vt_linkup;
 	}
-- 
2.34.1

